runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -0,0 +1,438 @@
+
+/*******************************************************************************
+ * Copyright (c) 2021, 2021 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "optimizer/TreeLowering.hpp"
+
+#include "compile/Compilation.hpp"
+#include "compile/SymbolReferenceTable.hpp"
+#include "il/Block.hpp"
+#include "il/Block_inlines.hpp"
+#include "infra/ILWalk.hpp"
+
+const char *
+TR::TreeLowering::optDetailString() const throw()
+   {
+   return "O^O TREE LOWERING: ";
+   }
+
+int32_t
+TR::TreeLowering::perform()
+   {
+
+   TR::ResolvedMethodSymbol* methodSymbol = comp()->getMethodSymbol();
+   for (TR::PreorderNodeIterator nodeIter(methodSymbol->getFirstTreeTop(), comp()); nodeIter != NULL ; ++nodeIter)
+      {
+      TR::Node* node = nodeIter.currentNode();
+      TR::TreeTop* tt = nodeIter.currentTree();
+
+      if (TR::Compiler->om.areValueTypesEnabled())
+         {
+         lowerValueTypeOperations(node, tt);
+         }
+      }
+
+   return 0;
+   }
```

Comment:
On a more general note, at the moment this opt pass is only required for VT enabled builds and it is specifically looking to transform trees of interest to VT.  I'm wondering if this is needlessly adding overhead to all JDK builds that don't support VT (i.e., JDK8, 11, 16) by doing a pre-order node traversal that will never find anything of interest.

And currently, even for VT builds, the nodes of interest are just below the treetops so a full pre-order node traversal isn't really required (is it?).  Perhaps we should limit this to just a treetop iteration to start with.

Hence, I would be in favour of enabling this optimization only in VT-enabled builds.  Should there be a need to add a new lowering operation as part of this framework in the future, we can consider how best to enable it to maximize performance.

Reply:
@liqunl since you expressed interest in using this opt as well, do you have any opinions about this?

Reply:
Until there is another use for this opt pass (and it is perfectly fine that other transformations would piggyback on it) I would say only perform it in VT-enabled builds.  That way, builds for which VT won't apply (JDK 8, 11, 17) won't be impacted by it.

Should some other transformation come along later that would be applicable to all builds (i.e., JDK8, 11, 17) then we should enable it more broadly then.

As for the walk itself, I'm fine leaving it as a pre-order node traversal if there is another imminent use case for MH.  @liqunl ?

Reply:
I've added a check in https://github.com/eclipse/openj9/pull/11968/commits/6495f850594ff714c50eafd542aeb673b3fd191a to skip this opt in non-VT builds. I'm leaving the check inside the loop so that only one place needs to be changed to re-enable the opt.